operators/src/engine/execution_context.rs

```diff
@@ -1,46 +1,163 @@
-use geoengine_datatypes::{
-    primitives::Coordinate2D,
-    raster::{GridShape2D, TilingSpecification},
-};
-
-use crate::concurrency::{ThreadPool, ThreadPoolContext};
+use crate::concurrency::ThreadPool;
+use crate::engine::{QueryRectangle, VectorResultDescriptor};
+use crate::error::Error;
+use crate::mock::MockDataSetDataSourceLoadingInfo;
+use crate::source::OgrSourceDataset;
+use crate::util::Result;
+use geoengine_datatypes::dataset::DataSetId;
+use geoengine_datatypes::raster::GridShape;
+use geoengine_datatypes::raster::TilingSpecification;
+use std::any::Any;
+use std::collections::HashMap;
+use std::fmt::Debug;
 use std::path::PathBuf;
 
-/// A context that provides certain utility access during operator execution
-/// TODO: maybe this is rather a trait
-#[derive(Debug, Clone)]
-pub struct ExecutionContext<'pool> {
-    pub raster_data_root: PathBuf,
-    pub thread_pool: ThreadPoolContext<'pool>,
-    pub tiling_specification: TilingSpecification,
+/// A context that provides certain utility access during operator initialization
+pub trait ExecutionContext:
+    Send
+    + Sync
+    + LoadingInfoProvider<MockDataSetDataSourceLoadingInfo, VectorResultDescriptor>
+    + LoadingInfoProvider<OgrSourceDataset, VectorResultDescriptor>
+{
+    fn thread_pool(&self) -> &ThreadPool;
+    fn raster_data_root(&self) -> Result<PathBuf>; // TODO: remove once GdalSource uses LoadingInfo
+    fn tiling_specification(&self) -> TilingSpecification;
 }
 
-/// Create a provider for execution contexts in test environments
-pub struct MockExecutionContextCreator {
-    raster_data_root: PathBuf,
-    thread_pool: ThreadPool,
-    tiling_specification: TilingSpecification,
+pub trait LoadingInfoProvider<T, U> {
+    fn loading_info(&self, data_set: &DataSetId) -> Result<Box<dyn LoadingInfo<T, U>>>;
 }
 
-impl MockExecutionContextCreator {
-    pub fn context(&self) -> ExecutionContext {
-        ExecutionContext {
-            raster_data_root: self.raster_data_root.clone(),
-            thread_pool: self.thread_pool.create_context(),
-            tiling_specification: self.tiling_specification,
-        }
+pub trait LoadingInfo<T, U>: Debug + Send + Sync {
+    fn get(&self, query: QueryRectangle) -> Result<T>;
+    fn meta(&self) -> Result<U>;
+
+    fn box_clone(&self) -> Box<dyn LoadingInfo<T, U>>;
+}
+
+impl<T, U> Clone for Box<dyn LoadingInfo<T, U>> {
+    fn clone(&self) -> Box<dyn LoadingInfo<T, U>> {
+        self.box_clone()
     }
 }
 
-impl Default for MockExecutionContextCreator {
+pub struct MockExecutionContext {
+    pub raster_data_root: PathBuf,
```

Comment:
> ChristianBeilschmidt commented on 18.01.2021
Why do you make this pub when you provide an add method to add LoadingInfo?

it's only for mocking purposes so why not make it as flexible as possible?